crates/rolldown/src/ast_scanner/side_effect_detector/mod.rs

```diff
@@ -160,7 +181,13 @@ impl<'a> SideEffectDetector<'a> {
       }
       Expression::ClassExpression(cls) => self.detect_side_effect_of_class(cls),
       // Accessing global variables considered as side effect.
-      Expression::Identifier(ident) => self.is_unresolved_reference(ident),
+      Expression::Identifier(ident) => {
```

Comment:
global variable `undefined` is considered as side effect free. https://esbuild.github.io/try/#YgAwLjIwLjIALS1idW5kbGUgLS1taW5pZnk9ZmFsc2UgLS1mb3JtYXQ9ZXNtAGUAZW50cnkuanMAY2xhc3MgVCB7CiAgW3VuZGVmaW5lZF0oKSB7fQp9AABmaWxlLmpzAHZhciB2ID0gdHlwZW9mIGxvY2FsVmFyCm1vZHVsZS5leHBvcnRzID0gdg

Reply:
Should just use `detect_side_effect_of_indentifier` to unite the logic of detecting if a identifer have side effect. It's not just underfined but a list of identifiers.

---edited

It's better we just transform `member_expr.object` to expression and call `detect_side_effect_of_expression`.